modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/serialize/relation/RelGraph.java

```diff
@@ -17,10 +17,98 @@
 
 package org.apache.ignite.internal.processors.query.calcite.serialize.relation;
 
-import org.apache.ignite.internal.processors.query.calcite.serialize.Graph;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.calcite.linq4j.Ord;
+import org.apache.calcite.util.ImmutableIntList;
+import org.apache.ignite.internal.processors.query.calcite.util.Commons;
+import org.apache.ignite.internal.util.GridIntList;
 
 /**
  * RelNode tree intermediate representation. Used by communication layer to send a query plan to remote node.
  */
-public class RelGraph extends Graph<RelGraphNode> {
+public class RelGraph implements Serializable {
```

Comment:
Why do we use graph structure for storing tree-like query plans? Do we have a protection against circles in the graph?

Reply:
There is no cycles now, but such structure is needed for serialization, flat structures are better there (because it makes easier to implement custom serialisation protocol)

No, there is no protection, in case of a cycle you'll get StackOverflowException on any tree transfer operation (almost all of them are recursive)

Reply:
ok